### PR TITLE
v4 Read Throughput Perf improvement

### PR DESCRIPTION
**Pull request by moderakh**
### This PR contains multiple performance related improvements for v4 sdk (more PRs to come).
**overall this PR increases Single Point Read Throughput by 10%**

To review this PR please review each individual commit which captures a specific perf improvement aspect (This will be easier for you to review).

1) SessionContainer has multiple internal locking which cause lock contention. ae54205ad77ae63be9a2966a2b497c3f5a2859de improves contention locking in session container

2) AddressCache caches constant addresses. We should avoid parsing cached addresses to URI on every request, this 16b77b87e3ef5e83cf14f454b7a1809444dd4d64 improves Address Caches URI parsing.

3) for each request in multiple places we convert URI <-> URL back and forth which requires parsing and is expensive. this f1fe01e89796bc79a4a4a4f6647e4754881a88b5 eliminates URI <-> URL conversion.

4) String regex operation is expensive, e4a49f189a8a746fa24dd0292151162683aca5c6 replaces String regex with other more performant string operation.

#################
#################

**How throughput increase is measured:**

on an Ubuntu Linux VM 16CPU core against a 1M (1,000,000) throughput collection
run ReadThroughput benchmark with 2000 concurrency for reading 10M (10,000,000) documents.

Running time of 8 sample runs for reading 10M documents without this PR took:
264s, 270s, 268s, 263s, 263s, 266s, 268s, 269s

Running time of 8 sample runs for reading 10M documents with this PR improvements took:
244s, 242s, 244s, 243s, 240s, 243s, 243s, 240s

**Throughput (doc read per second) without this PR:**
(10,000,000/((264+270+268+263+263+266+268+269)/8)=**37,541** 

**Throughput (doc read per second) with this PR:**
(10,000,000/((244+242+244+243+240+243+243+240)/8)=**41,258**

**Throughput (doc read per second) increase:**
(41,258-37,541)/37,541=**9.9%**

### CODE REVIEW

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/SessionContainer.java [hunk 1]
```diff
@@ -219,29 +219,50 @@ private void setSessionToken(ResourceId resourceId, String collectionName, Strin
...
-                    if (existingTokens == null) {
-                        ConcurrentHashMap<String, ISessionToken> tokens = new ConcurrentHashMap<>();
-                        tokens.put(partitionKeyRangeId, parsedSessionToken);
-                        return tokens;
+        ConcurrentHashMap<String, ISessionToken> existingTokensIfAny = this.collectionResourceIdToSessionTokens.get(resourceId.getUniqueDocumentCollectionId());
+
+        if (existingTokensIfAny != null) {
+            // if an entry for this collection exists, no need to lock the outer ConcurrentHashMap.
+            ConcurrentHashMap<String, ISessionToken>  existingTokens = existingTokensIfAny;
+            existingTokens.merge(partitionKeyRangeId, parsedSessionToken, (existingSessionTokens, newSessionToken) -> {
```
Comment (kushagraThapar) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/SessionContainer.java]: Line 227 - 237 code is duplicated below in line 252 - 262.
To make the code better in readability, can we extract this piece of code in its own helper method ?

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/HttpUtils.java [hunk 2]
```diff
@@ -45,14 +45,14 @@ public static String urlDecode(String url) {
...
 
-    public static URI toURI(String uri) {
-        try {
-            return new URI(uri);
-        } catch (Exception e) {
-            log.error("failed to parse {}", uri, e);
-            throw new IllegalArgumentException("failed to parse uri " + uri, e);
-        }
-    }
+//    public static URI toURI(String uri) {
```
Comment (kushagraThapar) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/HttpUtils.java]: Do we need to keep this commented method ?

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/SessionTokenHelper.java [hunk 3]
```diff
@@ -57,8 +57,9 @@ public static void setPartitionLocalSessionToken(RxDocumentServiceRequest reques
         if (request.requestContext.sessionToken == null) {
             request.getHeaders().remove(HttpConstants.HttpHeaders.SESSION_TOKEN);
         } else {
+
             request.getHeaders().put(HttpConstants.HttpHeaders.SESSION_TOKEN,
-                                     String.format("%1s:%2s", partitionKeyRangeId, request.requestContext.sessionToken.convertToString()));
+                                     concatPartitionKeyRangeIdWithSessionToken(partitionKeyRangeId, request.requestContext.sessionToken.convertToString()));
```
Comment (kushagraThapar) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/SessionTokenHelper.java]: Glad we made this change. We should make sure to populate this change everywhere, specially in `OrderByDocumentQueryExecutionContext`.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/HttpUtils.java (hunk 2 above)
Comment (moderakh) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/HttpUtils.java]: no. will remove. thanks for pointing out.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/SessionTokenHelper.java (hunk 3 above)
Comment (moderakh) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/SessionTokenHelper.java]: The scope of current PR is only single point read. I will investigate the query separately.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/HttpUtils.java (hunk 2 above)
Comment (moderakh) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/HttpUtils.java]: addressed in 048bec85ce063b82dc76c07ac0b26a18e6b5e0dd

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/SessionContainer.java (hunk 1 above)
Comment (moderakh) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/SessionContainer.java]: good suggestion. Thanks. addressed in 048bec85ce063b82dc76c07ac0b26a18e6b5e0dd

**Pull request merged**